File: 1330-longest-arithmetic-subsequence-of-given-difference/1330-longest-arithmetic-subsequence-of-given-difference.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int diff) {
        map<int, int> dp;
        int res = 1;
        
        for(int i = 0; i < arr.size(); i++){
            if(dp.count(arr[i] - diff)){
                dp[arr[i]] = 1 + dp[arr[i] - diff];
            }
            else{
                dp[arr[i]] = 1;
            }
            res = max(res, dp[arr[i]]);
        }

        return res;
    }
};

File: 1330-longest-arithmetic-subsequence-of-given-difference/1330-longest-arithmetic-subsequence-of-given-difference.py
class Solution:
    def longestSubsequence(self, arr: List[int], diff: int) -> int:
        dp = {}
        n = len(arr)
        res = 1
        for num in arr:
            if (num - diff) in dp:
                dp[num] = dp[num - diff] + 1
            else:
                dp[num] = 1

            res = max(res, dp[num])

        return res
